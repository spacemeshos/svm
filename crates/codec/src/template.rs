//! Encoding for binary [`Template`]s.
//!
//!  [`Template`] Binary Format
//!
//!  Important: There are no assumptions regarding the order of the `Section`s
//!
//!
//! ```text
//!
//!  +----------------+
//!  |                |
//!  |   #Sections    |
//!  |                |
//!  +----------------+
//!  |                |
//!  |  Code Section  |
//!  |                |
//!  +----------------+
//!  |                |
//!  | Data Section   |
//!  |                |
//!  +----------------+
//!  |                |
//!  | Ctors Section  |
//!  |                |
//!  +----------------+
//!  |                |
//!  | Header Section | (Optional)
//!  |                |
//!  +----------------+
//!  |                |
//!  | Schema Section | (Optional)
//!  |                |
//!  +----------------+
//!  |                |
//!  |  API Section   | (Optional)
//!  |                |
//!  +----------------+
//!  |                |
//!  | Deploy Section | (Optional, will be derived from the `Transaction Envelope` and `Transaction Context`)
//!  |                |
//!  +----------------+
//!
//! ```

use std::collections::HashSet;
use std::io::Cursor;

use crate::section::decode::decode_sections;
use crate::section::SectionsEncoder;
use crate::ParseError;
use svm_types::{SectionKind, Template};

/// Encodes a `Template` into binary
///
/// This is essentially equivalent to encoding the `Sections` residing within the `Template`
pub fn encode(template: &Template) -> Vec<u8> {
    let sections = template.sections();

    let mut encoder = SectionsEncoder::with_capacity(sections.len());
    encoder.encode(&sections);

    let bytes = encoder.finish();

    bytes
}

/// Decodes a list of `Section`s that we're interested at (see `interest`
/// parameter) and returns them wrapped within a `Template`.
///
/// If the input `interests` is `None` - decodes any kind `Section` belonging to
/// the `Template` pointed by the input `cursor`.
pub fn decode(
    cursor: Cursor<&[u8]>,
    interests: Option<HashSet<SectionKind>>,
) -> Result<Template, ParseError> {
    let sections = decode_sections(cursor, interests)?;

    let template = Template::from_sections(sections);
    Ok(template)
}

#[cfg(test)]
mod tests {
    use super::*;

    use maplit::hashset;

    use svm_layout::{FixedLayout, Id, Layout, RawVar};
    use svm_types::{
        Address, CodeKind, CodeSection, CtorsSection, DataSection, DeploySection, GasMode,
        HeaderSection, Layer, TemplateAddr, TransactionId,
    };

    fn make_code_section() -> CodeSection {
        CodeSection::new(CodeKind::Wasm, vec![0xC0, 0xDE], 0x01, GasMode::Fixed, 1)
    }

    fn make_data_section() -> DataSection {
        let mut section = DataSection::default();

        let var1 = RawVar::new(Id(0), 0, 10);
        let var2 = RawVar::new(Id(1), 10, 5);

        let fixed = FixedLayout::new(vec![var1, var2]);
        section.add_layout(Layout::Fixed(fixed));

        section
    }

    fn make_ctors_section() -> CtorsSection {
        let mut section = CtorsSection::default();

        section.push("init".to_string());

        section
    }

    fn make_header_section() -> HeaderSection {
        HeaderSection::new(2, "My Template".to_string(), "A few words".to_string())
    }

    fn make_deploy_section() -> DeploySection {
        let tx_id = TransactionId::repeat(0xFF);
        let layer = Layer(10);
        let deployer = Address::repeat(0xAB);
        let template = TemplateAddr::repeat(0xCD);

        DeploySection::new(tx_id, layer, deployer, template)
    }

    #[test]
    fn encode_template() {
        let code = make_code_section();
        let data = make_data_section();
        let ctors = make_ctors_section();
        let header = make_header_section();

        let mut template = Template::new(code, data, ctors).with_header(header);

        let deploy = make_deploy_section();
        template.deploy_section = Some(deploy);

        let bytes = encode(&template);

        let cursor = Cursor::new(&bytes[..]);

        let interests = hashset! {
            SectionKind::Code,
            SectionKind::Data,
            SectionKind::Ctors,
            SectionKind::Header,
            SectionKind::Deploy
        };

        let sections = decode_sections(cursor, Some(interests)).unwrap();

        assert_eq!(template.sections(), sections);
    }
}
