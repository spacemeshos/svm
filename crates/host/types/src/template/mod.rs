//! Generally, SVM needs to retrieves different pieces of data in different contexts.
//! We can roughly categorize the content as Core and Non-Core.
//!
//! It's pivotal to be able to fetch only the required data as fast as we can.
//! Secondly, it's crucial to make the format of a Template easy to evolve in the future.
//!
//! The solution introduced in this PR is called the Template Sections.
//! Instead of looking at a Template as a single unit - from now on, we'll consider a Template
//! as a collection of independent Section. Each Section will serve as single purpose and belong to
//! a either the Core or Non-Core category.
//!
//! The Template is an evolving entity. There is the pre-deploy stage and post-deploy stages.
//! Instead of having something such as PreDeployTemplate and PostDeployTemplate in the code base - deploying a Template means adding a Deploy Section to a collection of Section.
//!
//! If in the future we'll add a removal feature for a Template - we could come up with a new Removal Section or similar.
//!
//! Moving into the Sections design, we can now go further and store groups of Sections in different indexes very naturally.
//! Upon fetch, we should query the required indexes and get the corresponding Sections back.
//!
//! This PR also implements a skip Section functionality when loading data.
//!
//! Each Section is prefixed with a Section Preview, so when decoding the raw fetched data, we can ask to skip Sections that we don't want and only decode the ones we want.

mod api;
mod code;
mod ctors;
mod data;
mod deploy;
mod header;
mod schema;
mod section;

pub use api::ApiSection;
pub use code::{CodeKind, CodeSection};
pub use ctors::CtorsSection;
pub use data::DataSection;
pub use deploy::DeploySection;
pub use header::HeaderSection;
pub use schema::SchemaSection;
pub use section::{Section, SectionKind, SectionLike, Sections, SectionsIter};

use svm_hash::{Blake3Hasher, Hasher};
use svm_layout::FixedLayout;

use crate::{BytesPrimitive, TemplateAddr};

/// An in-memory representation of a `Template`
#[allow(missing_docs)]
#[derive(Debug, Clone, PartialEq)]
pub struct Template {
    sections: Sections,
}

impl Template {
    /// Creates a new [`Template`] only with the mandatory sections.
    pub fn new(
        code_section: CodeSection,
        data_section: DataSection,
        ctors_section: CtorsSection,
    ) -> Self {
        let mut sections = Sections::default();
        sections.insert(Section::Code(code_section));
        sections.insert(Section::Data(data_section));
        sections.insert(Section::Ctors(ctors_section));

        Self::from_sections(sections)
    }

    /// Creates a new `Template`
    ///
    /// # Panics
    ///
    /// Panics if and only if `sections` does *not* contain all mandatory
    /// [`Template`] sections.
    pub fn from_sections(sections: Sections) -> Self {
        Self { sections }
    }

    /// Adds, removes or replaces a [`HeaderSection`] in `self`.
    pub fn with_header(mut self, header: Option<HeaderSection>) -> Self {
        if let Some(header) = header {
            self.sections.insert(Section::Header(header));
        }

        self
    }

    /// Adds, removes or replaces a [`DeploySection`] in `self`.
    pub fn with_deploy(mut self, deploy: Option<DeploySection>) -> Self {
        if let Some(deploy) = deploy {
            self.sections.insert(Section::Deploy(deploy));
        }

        self
    }

    /// Borrows the `Sections` of the `Template`
    pub fn sections(&self) -> &Sections {
        &self.sections
    }

    /// Borrows the `Header Section`
    ///
    /// # Panics
    ///
    /// Panics if there is no `Header Section`
    pub fn header_section(&self) -> &HeaderSection {
        let section = self.get(SectionKind::Header);
        section.as_header()
    }

    /// Borrows the `Code Section`
    ///
    /// # Panics
    ///
    /// Panics if there is no `Code Section`
    pub fn code_section(&self) -> &CodeSection {
        let section = self.get(SectionKind::Code);
        section.as_code()
    }

    /// Borrows the `Template's` code (a blob)
    ///
    /// # Panics
    ///
    /// Panics if there is no `Code Section`
    pub fn code(&self) -> &[u8] {
        let section = self.code_section();
        section.code()
    }

    /// Borrows the `Data Section`
    ///
    /// # Panics
    ///
    /// Panics if there is no `Data Section`
    pub fn data_section(&self) -> &DataSection {
        let section = self.get(SectionKind::Data);
        section.as_data()
    }

    /// Returns an immutable borrow of `self`'s [`FixedLayout`].
    pub fn fixed_layout(&self) -> &FixedLayout {
        // See <https://github.com/spacemeshos/svm/issues/281>.
        let data = self.data_section();
        let layouts = data.layouts();

        let layout = layouts.first().unwrap();
        layout.as_fixed()
    }

    /// Borrows the `Ctors Section`
    ///
    /// # Panics
    ///
    /// Panics if there is no `Ctors Section`
    pub fn ctors_section(&self) -> &CtorsSection {
        let section = self.get(SectionKind::Ctors);
        section.as_ctors()
    }

    /// Returns the `Ctors` of the `Template`
    ///
    /// # Panics
    ///
    /// Panics if there is no `Ctors Section`
    pub fn ctors(&self) -> &[String] {
        let section = self.ctors_section();
        section.ctors()
    }

    /// Returns whether `Template` has a `ctor` equals `func_name`
    ///
    /// # Panics
    ///
    /// Panics if there is no `Ctors Section`
    pub fn is_ctor(&self, func_name: &str) -> bool {
        let ctors = self.ctors();
        ctors.iter().any(|ctor| ctor.as_str() == func_name)
    }

    /// Borrows the `Schema Section`
    ///
    /// # Panics
    ///
    /// Panics if there is no `Schema Section`
    pub fn schema_section(&self) -> &SchemaSection {
        let section = self.get(SectionKind::Schema);
        section.as_schema()
    }

    /// Sets the `DeploySection` to a `Template`
    pub fn set_deploy_section(&mut self, section: DeploySection) {
        debug_assert!(self.sections.contains(SectionKind::Deploy) == false);

        self.sections.insert(section.into());
    }

    /// Borrows the `Deploy Section`
    ///
    /// # Panics
    ///
    /// Panics if there is no `Deploy Section`
    pub fn deploy_section(&self) -> &DeploySection {
        let section = self.get(SectionKind::Deploy);
        section.as_deploy()
    }

    /// Returns the `Address` of a deployed `Template`
    ///
    /// # Panics
    ///
    /// Panics if there is no `Deploy Section`
    pub fn template_addr(&self) -> &TemplateAddr {
        let section = self.deploy_section();
        section.template()
    }

    /// Borrows the `Section` of the requested `SectionKind`
    ///
    /// # Panics
    ///
    /// Panics if there is no `Section` having the requested `SectionKind`
    pub fn get(&self, kind: SectionKind) -> &Section {
        self.sections.get(kind)
    }

    /// Borrows the `Section` of the requested `SectionKind`
    ///
    /// Returns `None` when there is no `Section` of the specified `SectionKind`
    pub fn try_get(&self, kind: SectionKind) -> Option<&Section> {
        self.sections.try_get(kind)
    }

    /// Returns whether Section exists
    pub fn contains(&self, kind: SectionKind) -> bool {
        self.sections.try_get(kind).is_some()
    }
}

/// Calculates the address of a newly deployed [`Template`].
pub fn compute_template_addr(code_section: &CodeSection) -> TemplateAddr {
    let hash = Blake3Hasher::hash(code_section.code());

    TemplateAddr::new(&hash[..TemplateAddr::N])
}
