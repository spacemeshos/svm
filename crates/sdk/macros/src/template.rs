use proc_macro2::{Ident, Span, TokenStream};
use quote::quote;

use serde_json::Value;

use syn::{Error, Item, ItemMod, ItemStruct, ItemType, ItemUse, Result};

use super::{function, r#struct};
use crate::{schema, Function, Schema, Struct};

#[cfg(feature = "api")]
use crate::api;

use r#function::{func_attrs, has_default_fundable_hook_attr};
use r#struct::has_storage_attr;

pub struct Template {
    name: Ident,
    functions: Vec<Function>,
    structs: Vec<Struct>,
    imports: Vec<ItemUse>,
    aliases: Vec<ItemType>,
    default_fundable_hook: Option<Ident>,
}

impl Template {
    pub fn name(&self) -> String {
        self.name.to_string()
    }

    pub fn functions(&self) -> &[Function] {
        &self.functions
    }

    pub fn structs(&self) -> &[Struct] {
        &self.structs
    }

    pub fn imports(&self) -> &[ItemUse] {
        &self.imports
    }

    pub fn aliases(&self) -> &[ItemType] {
        &self.aliases
    }

    pub fn default_fundable_hook(&self) -> Option<Ident> {
        self.default_fundable_hook.clone()
    }

    pub fn set_default_fundable_hook(&mut self, hook: Ident) {
        self.default_fundable_hook = Some(hook)
    }
}

pub fn expand(_args: TokenStream, input: TokenStream) -> Result<(Schema, TokenStream)> {
    let module = syn::parse2(input)?;
    let template = parse_template(module)?;
    let schema = schema::template_schema(&template)?;

    let imports = template.imports();
    let aliases = template.aliases();

    let structs = expand_structs(&template)?;
    let functions = expand_functions(&template)?;
    let alloc_func = alloc_func_ast();

    #[cfg(feature = "api")]
    let api = api::json_api(&schema);

    #[cfg(feature = "api")]
    let stream = api::json_tokenstream(&api);

    #[cfg(not(feature = "api"))]
    let stream = quote! { "" };

    #[cfg(feature = "api")]
    let data = api::json_data_layout(&schema);

    #[cfg(feature = "api")]
    write_schema(&template, &api, &data);

    let ast = quote! {
        // #(#imports)*

        // #(#aliases)*

        #alloc_func

        #structs

        #functions

        #[cfg(all(feature = "api", not(target_arch = "wasm32")))]
        pub fn raw_schema() -> String {
            #stream.to_string()
        }
    };

    Ok((schema, ast))
}

pub fn parse_template(mut raw_template: ItemMod) -> Result<Template> {
    let name = raw_template.ident.clone();

    let mut functions = Vec::new();
    let mut structs = Vec::new();
    let mut imports = Vec::new();
    let mut aliases = Vec::new();

    let (_, content) = raw_template.content.take().unwrap();

    for item in content {
        // TODO: Is is possible to extract the `item` real `Span`?
        let span = Span::call_site();

        match item {
            Item::Fn(item) => {
                let func = Function::new(item, functions.len());
                functions.push(func);
            }
            Item::Struct(item) => {
                let strukt = Struct::new(item);
                structs.push(strukt);
            }
            Item::Use(item) => imports.push(item),
            Item::Type(item) => aliases.push(item),
            Item::Const(..) => {
                let msg = "declaring `const` inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Enum(..) => {
                let msg = "declaring `enum` inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::ExternCrate(..) => {
                let msg = "using `extern crate` inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::ForeignMod(..) => {
                let msg =
                    "using foreign items such as `extern \"C\"` inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Impl(..) => {
                let msg = "using `impl` inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Macro(..) => {
                let msg = "declaring `macro_rules!` inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Macro2(..) => {
                let msg = "declaring `macro` inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Mod(..) => {
                let msg = "declaring new modules inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Static(..) => {
                let msg = "declaring new `static` items inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Trait(..) => {
                let msg = "declaring new traits inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::TraitAlias(..) => {
                let msg = "using trait aliases inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Union(..) => {
                let msg = "declaring `union` inside `#[template]` is not supported.";
                return Err(Error::new(span, msg));
            }
            Item::Verbatim(item) => {
                let msg = format!("invalid Rust code: {}", item);
                return Err(Error::new(span, msg));
            }
            Item::__TestExhaustive(..) => unreachable!(),
        }
    }

    let mut template = Template {
        name,
        functions,
        structs,
        imports,
        aliases,
        default_fundable_hook: None,
    };

    let default = extract_default_fundable_hook(&template)?;

    if default.is_some() {
        template.set_default_fundable_hook(default.unwrap());
    }

    Ok(template)
}

fn extract_default_fundable_hook(template: &Template) -> Result<Option<Ident>> {
    let span = Span::call_site();
    let mut seen_default_fundable_hook = false;
    let mut default = None;

    for func in template.functions().iter() {
        let attrs = func_attrs(func).unwrap();

        if has_default_fundable_hook_attr(&attrs) {
            if seen_default_fundable_hook {
                return Err(Error::new(
                    span,
                    "There can be only a single default `fundable hook`",
                ));
            }

            seen_default_fundable_hook = true;

            default = Some(func.raw_name());
        }
    }

    Ok(default)
}

#[cfg(all(feature = "api", target_arch = "wasm32"))]
fn write_schema(template: &Template, api: &Value, data: &Value) {
    api::json_write(&format!("{}-api.json", template.name()), api);
    api::json_write(&format!("{}-data.json", template.name()), data);
}

#[cfg(any(not(feature = "api"), not(target_arch = "wasm32")))]
fn write_schema(template: &Template, api: &Value, data: &Value) {
    //
}

fn expand_structs(template: &Template) -> Result<TokenStream> {
    let mut structs = Vec::new();

    validate_structs(template)?;

    for strukt in template.structs() {
        let strukt = r#struct::expand(strukt)?;

        structs.push(strukt);
    }

    let ast = quote! {
        #(#structs)*
    };

    Ok(ast)
}

fn validate_structs(template: &Template) -> Result<()> {
    let mut seen_storage = false;

    for strukt in template.structs() {
        match strukt.attrs() {
            Ok(attrs) => {
                if has_storage_attr(attrs) {
                    if seen_storage {
                        let msg = format!("A Template can have only a single `#[storage]`");
                        let span = Span::call_site();

                        return Err(Error::new(span, msg));
                    }

                    seen_storage = true;
                }
            }
            Err(err) => return Err(err.clone()),
        }
    }

    Ok(())
}

fn expand_functions(template: &Template) -> Result<TokenStream> {
    validate_funcs(template)?;

    let mut funcs = Vec::new();

    for func in template.functions() {
        let func = function::expand(func, template)?;

        funcs.push(func);
    }

    let implicit_fundable_hook = if template.default_fundable_hook().is_some() {
        quote! {}
    } else {
        function::fundable_hook::expand_default()?
    };

    let ast = quote! {
        #(#funcs)*

        #implicit_fundable_hook
    };

    Ok(ast)
}

fn validate_funcs(template: &Template) -> Result<()> {
    Ok(())
}

fn alloc_func_ast() -> TokenStream {
    quote! {
        extern crate svm_sdk;

        #[no_mangle]
        pub extern "C" fn svm_alloc(size: u32) -> u32 {
            let ptr = svm_sdk::alloc(size as usize);

            ptr.offset() as u32
        }
    }
}
