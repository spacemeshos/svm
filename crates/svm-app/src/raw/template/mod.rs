//!                  `AppTemplate` Raw Format
//!  +-----------------------------------------------------+
//!  |            |                |                       |
//!  |  version   |  name length   |         name          |
//!  |    (a)     |   (varuint14)  |        (UTF-8)        |
//!  +____________|________________|_______________________+
//!  |               |                                     |
//!  |  #app-pages   |          Code #bytes                |
//!  | (`varuint14`) |           (4 bytes)                 |
//!  +_______________|_____________________________________+
//!  |                                                     |
//!  |                   Code (wasm)                       |
//!  +_____________________________________________________+
//!
//!
//!
//! (a) Proto Version Encoding
//! ===========================
//!
//!   ___________________________________________________
//!  |  MSB  |  non-MSB  |         Meaning              |
//!  ---------------------------------------------------
//!  |   1   |  x  x  x  |  Next nibble is relevant too |
//!  |   0   |  x  x  x  |  Next nibble isn't relevant  |
//!  |--------------------------------------------------|
//!
//!  The protocol `#bits` will be a multiplication of 3,
//!  and is encoded in Little-endian as an unsigned-integer.
//!

mod serialize;
mod validate;
mod wire;

pub use serialize::{DefaultAppTemplateDeserializer, DefaultAppTemplateSerializer};
pub use validate::validate_template;
pub use wire::{decode_deploy_template, encode_deploy_template};
