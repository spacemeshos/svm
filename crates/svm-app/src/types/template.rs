use std::fmt;

use serde::{Deserialize, Serialize};

use svm_common::Address;

/// We first parse the raw `app-template` bytes into an `AppTemplate` instance.
/// At that stage we don't know the `app-template` future `Address` yet.
#[allow(missing_docs)]
#[derive(Serialize, Deserialize, PartialEq)]
pub struct AppTemplate {
    pub name: String,
    pub author: Address,
    pub pages_count: u16,
    pub code: Vec<u8>,
}

impl fmt::Debug for AppTemplate {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let name = self.fmt_name(&self.name);
        let author = self.fmt_author(&self.author);
        let pages = self.fmt_pages_count(self.pages_count);
        let code = self.fmt_code(&self.code);

        let msg = [author, name, code, pages].join("\n");
        write!(f, "{}", msg)
    }
}

impl AppTemplate {
    fn fmt_name(&self, name: &str) -> String {
        format!("Name: {:?}", name)
    }

    fn fmt_author(&self, author: &Address) -> String {
        format!("Author: {:?}...", &author.as_slice()[0..8])
    }

    fn fmt_pages_count(&self, pages_count: u16) -> String {
        format!("#Pages: {:?}", pages_count)
    }

    fn fmt_code(&self, code: &[u8]) -> String {
        format!("Code: {:?}", &code[0..4])
    }
}
