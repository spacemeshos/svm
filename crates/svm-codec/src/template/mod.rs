//! `AppTemplate` Raw Format
//!  +-----------------------------------------------------+
//!  |            |                |                       |
//!  |  version   |  name length   |         name          |
//!  |    (a)     |  (varuint14)   |        (UTF-8)        |
//!  +____________|________________|_______________________+
//!  |               |                                     |
//!  |  Code #bytes  |          Code (WASM)                |
//!  |   (4 bytes)   |                                     |
//!  +_______________|_____________________________________+
//!  |               |             |         |             |
//!  |  Data-Layout  |  var #0     |         |   var #N    |
//!  |  #variables   |  length     |  . . .  |   length    |
//!  |  (varuint14)  | (varuint14) |         | (varuint14) |
//!  +_______________|_____________|_________|_____________+
//!
//!
//!
//! (a) Proto Version Encoding
//! ===========================
//!
//!   ___________________________________________________
//!  |  MSB  |  non-MSB  |         Meaning              |
//!  ---------------------------------------------------
//!  |   1   |  x  x  x  |  Next nibble is relevant too |
//!  |   0   |  x  x  x  |  Next nibble isn't relevant  |
//!  |--------------------------------------------------|
//!
//!  The protocol `#bits` will be a multiplication of 3,
//!  and is encoded in Little-endian as an unsigned-integer.
//!

mod serialize;
mod validate;
mod wire;

pub use serialize::{DefaultAppTemplateDeserializer, DefaultAppTemplateSerializer};
pub use validate::validate_template;
pub use wire::{decode_deploy_template, encode_deploy_template};
