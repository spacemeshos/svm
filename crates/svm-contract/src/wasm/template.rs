use serde::{Deserialize, Serialize};

use svm_common::Address;

/// We first parse the on-the-wire `app-template` deploy transaction into a `AppTemplate` instance.
/// At that stage we don't know the `app-template` future `Address` yet.
#[allow(missing_docs)]
#[derive(Serialize, Deserialize, PartialEq)]
pub struct AppTemplate {
    pub code: Vec<u8>,
    pub name: String,
    pub author: Address,
}

impl std::fmt::Debug for AppTemplate {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        let author = self.fmt_author(&self.author);
        let code = self.fmt_code(&self.code);

        let msg = [author, code].join("\n");

        write!(f, "{}", msg)
    }
}

impl AppTemplate {
    fn fmt_author(&self, author: &Address) -> String {
        format!("Author: {:?}...", &author.as_slice()[0..8])
    }

    fn fmt_code(&self, code: &[u8]) -> String {
        format!("Code: {:?}", &code[0..4])
    }
}
