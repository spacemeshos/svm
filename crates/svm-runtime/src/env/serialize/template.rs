use std::io::Cursor;

use svm_codec::template;
use svm_codec::{Field, ReadExt, WriteExt};

use svm_types::{AuthorAddr, Template};

use crate::env::traits;
use crate::env::ExtTemplate;

use traits::{TemplateDeserializer, TemplateSerializer};

/// `Template` default Serializer
pub struct DefaultTemplateSerializer;

/// `Template` default Deserializer
pub struct DefaultTemplateDeserializer;

impl TemplateSerializer for DefaultTemplateSerializer {
    fn serialize(template: &ExtTemplate) -> Vec<u8> {
        let mut w = Vec::new();

        let base = template.base();
        let author = template.author();

        template::encode_deploy_template(base, &mut w);

        w.write_address(author.inner());

        w
    }
}

impl TemplateDeserializer for DefaultTemplateDeserializer {
    fn deserialize(bytes: &[u8]) -> Option<ExtTemplate> {
        let mut cursor = Cursor::new(bytes);

        let base = match template::decode_deploy_template(&mut cursor) {
            Ok(base) => base,
            _ => return None,
        };

        let author = match cursor.read_address() {
            Ok(addr) => AuthorAddr::new(addr.into()),
            _ => return None,
        };

        let template = ExtTemplate::new(base, &author);

        Some(template)
    }
}

#[cfg(test)]
mod tests {
    use std::vec;

    use super::*;

    use svm_types::Address;

    use DefaultTemplateDeserializer as D;
    use DefaultTemplateSerializer as S;

    #[test]
    fn serialize_template() {
        let base = Template {
            version: 0,
            name: "My Template".to_string(),
            code: vec![0x0C, 0x00, 0x0D, 0x0E],
            layout: vec![10, 20, 30].into(),
            ctors: vec!["init".into(), "start".into()],
        };

        let author = Address::of("@author").into();
        let template = ExtTemplate::new(base, &author);

        let bytes = S::serialize(&template);

        let decoded = D::deserialize(&bytes[..]).unwrap();
        assert_eq!(decoded, template);
    }
}
