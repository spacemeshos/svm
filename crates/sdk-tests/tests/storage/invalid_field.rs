use svm_sdk::template;

#[template]
mod Template {
    #[storage]
    struct Storage {
        name: String,
    }
}

fn main() {}
