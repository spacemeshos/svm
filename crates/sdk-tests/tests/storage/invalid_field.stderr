error: custom attribute panicked
 --> $DIR/invalid_field.rs:3:1
  |
3 | #[template]
  | ^^^^^^^^^^^
  |
  = help: message: called `Result::unwrap()` on an `Err` value: Error("Invalid `#[storage]` field type: String")
